compiler/qsc_rca/src/common.rs

```diff
@@ -0,0 +1,354 @@
+    Discard,
+    Node(NodeId),
+    Tuple(Vec<PatId>),
+}
+
+/// Creates a vector of flattened input pattern elements.
+fn derive_callable_input_pattern_elements(
+    callable: &CallableDecl,
+    pats: &IndexMap<PatId, Pat>,
+) -> Vec<InputPatternElement> {
+    fn create_input_elements(
+        pat_id: PatId,
+        pats: &IndexMap<PatId, Pat>,
+    ) -> Vec<InputPatternElement> {
+        let pat = pats.get(pat_id).expect("pattern should exist");
+        match &pat.kind {
+            PatKind::Bind(ident) => {
+                vec![InputPatternElement {
+                    pat: pat_id,
+                    ty: pat.ty.clone(),
+                    kind: InputPatternElementKind::Node(ident.id),
+                }]
+            }
+            PatKind::Tuple(tuple_pats) => {
+                let mut tuple_params = vec![InputPatternElement {
+                    pat: pat_id,
+                    ty: pat.ty.clone(),
+                    kind: InputPatternElementKind::Tuple(tuple_pats.clone()),
+                }];
+                for tuple_item_pat_id in tuple_pats {
+                    let mut tuple_item_params = create_input_elements(*tuple_item_pat_id, pats);
+                    tuple_params.append(&mut tuple_item_params);
+                }
+                tuple_params
+            }
+            PatKind::Discard => vec![InputPatternElement {
+                pat: pat_id,
+                ty: pat.ty.clone(),
+                kind: InputPatternElementKind::Discard,
+            }],
+        }
+    }
+
+    create_input_elements(callable.input, pats)
+}
+
+#[derive(Debug)]
+pub struct Callee {
+    pub item: StoreItemId,
+    pub spec_functor: SpecFunctor,
+}
+
+#[derive(Debug, Default)]
+pub struct SpecFunctor {
+    pub adjoint: bool,
+    pub controlled: u8,
+}
+
+/// Tries to uniquely resolve the callable specialization referenced in a callee expression.
+pub fn try_resolve_callee(
+    expr_id: ExprId,
+    package_id: PackageId,
+    package: &impl PackageLookup,
+    locals_map: &impl LocalsLookup,
+) -> Option<Callee> {
+    // This is a best effort attempt to resolve a callee that currently only supports resolving
+    // global callables or locals that eventually resolve to global callables.
+    let expr = package.get_expr(expr_id);
+    match &expr.kind {
+        ExprKind::UnOp(operator, operand_expr_id) => {
+            try_resolve_un_op_callee(*operator, *operand_expr_id, package_id, package, locals_map)
+        }
+        ExprKind::Var(res, _) => match res {
+            Res::Item(item_id) => resolve_item_callee(package_id, *item_id),
+            Res::Local(node_id) => {
+                try_resolve_local_callee(*node_id, package_id, package, locals_map)
+            }
+            Res::Err => panic!("callee resolution should not be an error"),
+        },
+        // N.B. More complex callee expressions might require evaluation so we don't try to resolve them at compile
+        // time.
```

Comment:
All comments are “Nota Bene” otherwise they wouldn’t be there. Prefixing comments with "N.B." is just noise and reduces readiblity.

Reply:
Removed all "N.B." occurrences.